Add support for WebGL2 TexStorage2D

mmatyas opened the pull request:
Adds initial support for the WebGL2 `TexStorage2D` call, adds support for the related texture enums and enables some of the texture tests.

See: https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.6

<!-- Please describe your changes on the following line: -->

cc @jdm @zakorgy 

A couple of notes:

- Currently `tests/wpt/webgl/meta/conformance2/textures` is disabled for testing; this patch enables the `misc` subdirectory and disables the rest, which is why there are so many new test files. The primary test for `TexStorage2D` is `conformance2/textures/misc/tex-storage-2d.html`.
- Most of the test passes, except the "texSubImage2D should succeed on immutable texture as long as the format is compatible" checks. This produces INVALID_OPERATIONs [here](https://github.com/servo/servo/blob/d5e5414be372a28215c5162b1257765b89fbd619/components/script/dom/webglrenderingcontext.rs#L798) because the texture's image info is not set at that level. I'm probably forgetting to call `set_image_infos_at_level` somewhere but I'm not sure where.
- There's a duplication of the internal texture format list with eg. the renderbuffer code, I wonder if there would be a good common place for it.


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mmatyas committed:
Add support for WebGL2 TexStorage2D

Adds initial support for the WebGL2 `TexStorage2D` call, adds
support for the related texture enums and enables some of the
texture tests.

See: https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.6

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webgl2renderingcontext.rs, components/script/dom/webgltexture.rs, components/script/dom/webglrenderingcontext.rs
 * @kichjang: components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webgl2renderingcontext.rs, components/script/dom/webgltexture.rs, components/script/dom/webglrenderingcontext.rs

jdm commented on components/script/dom/webgltexture.rs:
```diff
@@ -416,6 +426,114 @@ impl WebGLTexture {
+        if self.is_immutable.get() {
+            return Err(WebGLError::InvalidOperation);
+        }
+
```
I believe we should call `self.initialize(..)` here like https://github.com/servo/servo/blob/2c77ab66d4e718437a236d8e48082ab670ea254f/components/script/dom/webglrenderingcontext.rs#L733-L745 . This is where the missing texture info comes from.

mmatyas commented:
Thanks, I was experimenting with adding a `self.initialize` step, but unfortunately I still run into the same invalid operation issue. I've included the changes as a separate work-in-progress commit, do you have some ideas about where it fails?

jdm commented on components/script/dom/webgltexture.rs:
```diff
@@ -416,6 +427,124 @@ impl WebGLTexture {
+            height as u32,
+            1,
+            TexFormat::from_gl_constant(internal_format).unwrap(),
+            levels as u32,
```
I stepped through the test failure in a debugger and realized that we are initializing a single level of the texture, but TexStorage2D expects all levels from 0 to levels to be initialized. See page 137 of https://www.khronos.org/registry/OpenGL/specs/es/3.0/es_spec_3.0.pdf#nameddest=section-3.8.4.

mmatyas replied:
Ah you're right! Indeed I should've read the spec a bit more closely, I'll do some restructuring now.

mmatyas commented:
So last week I've started restructuring things, which at first looked to be simple, but then I've ended up in the rabbit hole of dealing with texture formats: to store the format in the texture, I had to convert the integer to a TexFormat, but for that I had to add the enums for all the newly acceptable formats, but because the formats were now implemented, they were also accessible from each and every WebGL function, and so I also needed to implement the additional texture format and data type combination checks and update the verification of the texture related functions.

Anyway, there are still quite a few failing test cases, but here's the current WIP status of this patch. I've run a tidy on it but not updated or regenerated all the tests yet.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26135) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas commented:
Well it keeps getting more and more out of scope, so I'll try to stop here for now. I've cleaned up the debug code, added the tests and rebased to master. There might be still a few misses here and there, but things seem to work mostly fine.

jdm commented on components/canvas_traits/webgl.rs:
```diff
@@ -1018,32 +1100,257 @@ impl TexFormat {
+    }
+
+    pub fn usable_as_internal(self) -> bool {
+        /*match self {
```
We should remove this commented code.

jdm commented on components/script/dom/webgl_validations/tex_image_2d.rs:
```diff
@@ -657,3 +671,109 @@ impl<'a> WebGLValidator for CompressedTexSubImage2DValidator<'a> {
+        };
+        if !dimensions_valid {
+            context.webgl_error(InvalidEnum);
+            return Err(TexImageValidationError::InvalidBorder);
```
I think a bunch of these error values don't match what's actually happening.

mmatyas replied:
Whoops, indeed, fixed.

jdm reviewed (changes requested):
I agree, I think we should merge what we have before it keeps getting bigger. Thanks for all the work to fix up the missing pieces here!

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 1c79147 has been approved by `jdm`

<!-- @bors-servo r=jdm 1c7914783a6dbf0dca8f9bb536742ffbb5b3ba03 -->

bors-servo commented:
:hourglass: Testing commit 1c7914783a6dbf0dca8f9bb536742ffbb5b3ba03 with merge 19cbea94db5a56646a8d9b6eb3b8b8df5d114b6a...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ZaUxZEksQr-IHntWK-IPoA)

jdm commented:
Oof. Some more investigation is going to be needed here to figure out how to avoid these new webgl 1 failures, and deal with the remaining webgl 2 failures.
```
/_webgl/conformance/extensions/oes-texture-half-float-with-image-data.html
/_webgl/conformance/extensions/oes-texture-half-float-with-image.html
/_webgl/conformance/extensions/oes-texture-half-float-with-canvas.html
/_webgl/conformance/extensions/oes-texture-float-with-canvas.html
/_webgl/conformance/extensions/oes-texture-float-with-image-data.html
/_webgl/conformance/extensions/oes-texture-half-float-linear.html
/_webgl/conformance/extensions/oes-texture-float-with-image.html
/_webgl/conformance/extensions/oes-texture-float.html
/_webgl/conformance2/textures/misc/tex-storage-2d.html
/_webgl/conformance/extensions/oes-texture-float-linear.html
/_webgl/conformance2/textures/misc/tex-image-with-different-data-source.html
/_webgl/conformance2/textures/misc/tex-3d-mipmap-levels-intel-bug.html
/_webgl/conformance/extensions/oes-texture-half-float.html
/_webgl/conformance2/rendering/fs-color-type-mismatch-color-buffer-type.html

  ▶ Unexpected subtest result in /_webgl/conformance/state/gl-object-get-calls.html:
  │ FAIL [expected PASS] WebGL test #72: getTexParameter returned 10497 instead of null for invalid parameter enum: 0x8072
  │   → assert_true: getTexParameter returned 10497 instead of null for invalid parameter enum: 0x8072 expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1958:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  │ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  │ testInvalidArgument@http://web-platform.test:8000/_webgl/js/tests/gl-object-get-calls.js:45:19
  └ @http://web-platform.test:8000/_webgl/js/tests/gl-object-get-calls.js:491:20
```

mmatyas commented:
Ah yes that was caused by the new TexParameter types not having WebGL version checks, so they've got picked up in [this loop](https://github.com/servo/servo/blob/37023b24f256f7defd605e90e0873afee9197a1e/tests/wpt/webgl/tests/js/tests/gl-object-get-calls.js#L40). Adding the check didn't increase the patch too much, so I've just updated it with this fix.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit a77f0f9d5c1dc23277ddc722e9a32f76f235361f with merge 28f1c45765fce7f03ba0f9ecb115395f5721d7d2...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/LAWdXg8rT_aZ3lH0f9bktg)

jdm commented:
I'm tracking down the failures related to texture data types in webgl 1 tests.

jdm committed:
webgl: Fix support for float/half-float texture formats.

jdm commented:
tex-storage-2d.html fails intermittently because we don't initialize the new texture data storage to 0s like https://searchfox.org/mozilla-central/rev/9120151ddb35f2d4c37bfe613a54a4f10a9a3dc5/dom/canvas/WebGLTexture.cpp#526.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 1addd238ceee05d183bfd171a252da32704152ae with merge 6b27d91a10746cf51c285a8f1db1e3d912454792...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/X7jwhABbSDatbMgDOhXXAQ)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25853) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas committed:
webgl: Update texture test results

mmatyas commented:
Rebased to master and updated some of the tests (as a separate commit for now). I've noticed you added float formats to eg. RGBA, but with the current code these make some of the tests fail or crash at the moment ([example](https://github.com/servo/servo/blob/99cd30eaad786f7c0a113979e8a9f6b4d0250fec/tests/wpt/webgl/tests/conformance2/textures/misc/tex-image-with-bad-args.html#L66)). Should we skip them for now?

jdm commented:
No need to skip them if they're consistent results.

mmatyas commented:
Ok, updated the last commit with the remaining tests (`conformance/state/gl-object-get-call`, `conformance/textures/misc/texture-size-limit` and `conformance2/textures/misc/tex-image-with-bad-args`).

mmatyas committed:
webgl: Fix tidy

mmatyas commented:
Small update: I've fixed `GetTexParameter` in WebGL1 not returning invalid enum for WebGL2 parameters.

jdm committed:
webgl: Add missing RGB8 texture format combination.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26291) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas commented:
Rebased to master after the recent changes, tests seem to pass on my side as well. May I ask about your plans regarding this patch? If you eg. think it's too big, I could try splitting it into separate pull requests.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 3a8c86addfff1d89d86c5ad8c82d6858294df0de with merge d8f08e8506a40e7b8898ff785999b77a8ad22cb6...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QpXBGWGXRyGNAzz0JX2ipQ)

jdm commented:
> Rebased to master after the recent changes, tests seem to pass on my side as well. May I ask about your plans regarding this patch? If you eg. think it's too big, I could try splitting it into separate pull requests.

My plans for this patch are to get it merged by disabling any tests that are failing because of uninitialized texture data and otherwise updating the test results to match CI if they're consistent. I haven't had much time to make progress on this goal recently.

jdm committed:
Update test expectations.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit d72e719c316a94bf6e013f7c57325725f07ee186 with merge d373224ea73cd52587004a48c311653d7a69a704...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/IuSaxMa5S_KK5bg7czM9CQ)
State: approved= try=True

jdm commented:
@bors-servo try=mac-wpt retry

bors-servo commented:
:slightly_frowning_face: There is no try chooser mac-wpt for this repo, try one of: linux, mac, windows, windows-rdp, wpt, wpt-2020, wpt-mac, wpt-android, android, magicleap, arm

bors-servo commented:
:hourglass: Trying commit d72e719c316a94bf6e013f7c57325725f07ee186 with merge 1573c802b81095ad6e4f261b5c391ffdf06609ac...

jdm commented:
@bors-servo try=wpt-mac retry

bors-servo commented:
:hourglass: Trying commit d72e719c316a94bf6e013f7c57325725f07ee186 with merge d0a993caa41615deaa8e136dd8609fe42f19d8eb...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/fHfVVqjBTKenJn8iZvsENw)

jdm committed:
webgl: Return TEXTURE_IMMUTABLE_FORMAT as a boolean; don't panic on macOS for TEXTURE_IMMUTABLE_LEVELS.

jdm committed:
Update test results for macOS.

jdm commented:
@bors-servo try=wpt-mac

bors-servo commented:
:hourglass: Trying commit 16991460e020b78af4c2f1af756a8c2f443828e9 with merge 269e42914843d39d0e9f90291d8813bf000478dd...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QfsI5okJRKusWosP6gOdpQ)

jdm commented:
@bors-servo r+ try-

bors-servo commented:
:pushpin: Commit 1699146 has been approved by `jdm`

<!-- @bors-servo r=jdm 16991460e020b78af4c2f1af756a8c2f443828e9 -->

jdm commented:
@bors-servo r-

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 8d1b60e has been approved by `jdm`

<!-- @bors-servo r=jdm 8d1b60eb8979c67d62ec38b04508e09fc3f9c270 -->

bors-servo commented:
:hourglass: Testing commit 8d1b60eb8979c67d62ec38b04508e09fc3f9c270 with merge bc9f2879069b864112fbb898089d2719b3cb1d53...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/H2Yn1lhHRbCYQDDf1mxFXQ)

jdm commented:
@bors-serbo retry

bors-servo commented:
:hourglass: Testing commit 8d1b60eb8979c67d62ec38b04508e09fc3f9c270 with merge 34055bd9f91be17c210630e48e186317744fa1d3...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/aV4UL6qqQh-3d_csObR8EA)

jdm commented:
@bors-servo retry
* #25513

bors-servo commented:
:hourglass: Testing commit 8d1b60eb8979c67d62ec38b04508e09fc3f9c270 with merge 76701d4ccf12e8931160c2537c62a869c415155c...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/eUXZXsbqRMqzlNBqRaRcBg)

jdm commented:
@bors-servo retry
* #25513

bors-servo commented:
:hourglass: Testing commit 8d1b60eb8979c67d62ec38b04508e09fc3f9c270 with merge 4dacbcbf131d23b5b5d2d24cfd86dfc705b79129...

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 8d1b60eb8979c67d62ec38b04508e09fc3f9c270 with merge 462bd5dfa63017ee2035bd5dac54ae1b08d8f7d2...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QUyFCg-ISdu6orb6mYFmxA)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 8d1b60eb8979c67d62ec38b04508e09fc3f9c270 with merge da7dbc5533cbf2a726627360ddddcb527d2b8728...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/I7D_FWPYTbKOdJWiFvKi4Q)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26354) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 0c87590 has been approved by `jdm`

<!-- @bors-servo r=jdm 0c87590071d72f5722f1869e132cce3956c18faf -->

bors-servo commented:
:hourglass: Testing commit 0c87590071d72f5722f1869e132cce3956c18faf with merge e544a97ca66038b62daff4ed1bb0b4c4e16b9146...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/WYPi7yn-QxeisM4dvO8azA)

CYBAI commented:
```
  ▶ Unexpected subtest result in /_webgl/conformance/textures/misc/tex-sub-image-2d.html:
  │ FAIL [expected PASS] WebGL test #0: pixel at (1, 0) was (0, 0, 0), should be (1, 1, 1)
  │   → assert_true: pixel at (1, 0) was (0, 0, 0), should be (1, 1, 1) expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1988:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  │ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  │ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  └ @http://web-platform.test:8000/_webgl/conformance/textures/misc/tex-sub-image-2d.html:107:19
```

Not sure if this is an intermittent failure, other than it, other failures are #25513 and #23290.

jdm commented:
@bors-servo retry
* #26365 #25513 #23290

bors-servo commented:
:hourglass: Testing commit 0c87590071d72f5722f1869e132cce3956c18faf with merge 2b20f04e8b830b04b83ab3dab33471076a9e2731...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/DSMMrhg0T-ejCc_32EOeXA)
Approved by: jdm
Pushing 2b20f04e8b830b04b83ab3dab33471076a9e2731 to master...
